### ISSUE TITLE
Investigate strange speedup in santa tracker Android performance tests

### ISSUE THREAD

**Issue by wolfs** [gradle/gradle#15115]
When running a build with async-profiler attached, it looks like the build gets much faster towards the end ([build](https://builds.gradle.org/buildConfiguration/Gradle_Check_PerformanceTest14bucket1/39271118?buildTab=artifacts)):
![image](https://user-images.githubusercontent.com/423186/98706214-e2a13800-237e-11eb-8bf5-2ee86bde2f71.png)

We should find out what happens there. It may be that file system changes are not picked up any more.

---
cc: @gradle/execution

**Comment by wolfs** [comment]
I managed to reproduce the problem again, this time with a log ([build](https://builds.gradle.org/buildConfiguration/Gradle_Check_PerformanceTest14bucket1/39331424?buildTab=artifacts)).

The builds which are faster have the following output:
```
* Running measured build #38
VFS> Statistics since last build:
VFS> > Stat: Executed stat() x 0. getUnixMode() x 0
VFS> > FileHasher: Hashed 0 files (0 bytes)
VFS> > DirectorySnapshotter: Snapshot 0 directory hierarchies (visited 0 directories, 0 files and 0 failed files)
Received 0 file system events since last build while watching 1 hierarchies
Virtual file system retained information about 13520 files, 4736 directories and 94 missing files since last build
Configuration cache is an incubating feature.
Reusing configuration cache.

....

BUILD SUCCESSFUL in 185ms
189 actionable tasks: 189 up-to-date
Received 0 file system events during the current build while watching 1 hierarchies
Virtual file system retains information about 13520 files, 4736 directories and 94 missing files until next build
VFS> Statistics during current build:
VFS> > Stat: Executed stat() x 6. getUnixMode() x 0
VFS> > FileHasher: Hashed 0 files (0 bytes)
VFS> > DirectorySnapshotter: Snapshot 0 directory hierarchies (visited 0 directories, 0 files and 0 failed files)
```

It looks like the watching is stopped between builds and not restarted, at least we are not receiving any more events.

**Comment by wolfs** [comment]
I found this in the log:
```
Caught exception: Couldn't poll for events, error = 4
net.rubygrapefruit.platform.NativeException: Couldn't poll for events, error = 4
	at net.rubygrapefruit.platform.internal.jni.AbstractFileEventFunctions$NativeFileWatcher.executeRunLoop0(Native Method)
	at net.rubygrapefruit.platform.internal.jni.AbstractFileEventFunctions$NativeFileWatcher.access$100(AbstractFileEventFunctions.java:156)
	at net.rubygrapefruit.platform.internal.jni.AbstractFileEventFunctions$NativeFileWatcher$1.run(AbstractFileEventFunctions.java:169)
```

So the run loop did exit, and we didn't get any notification on the Gradle side, expecting the run loop to still work. I suppose we would receive a failure as soon as we try to stop watching something, though I am not sure.

**Comment by wolfs** [comment]
4 seems to be `EINTR`, so the system call was interrupted. Maybe we should retry on the native side in that occasion? We for sure need to propagate the problem to Gradle.

**Comment by wolfs** [comment]
I looked into the performance DB, and it seems like we only have these problems when running with async profiler attached.

**Comment by wolfs** [comment]
Fixed via #15160.
